metadata-ingestion/src/datahub/ingestion/source/vertexai.py

```diff
@@ -0,0 +1,795 @@
+            yield from self._get_training_job_workunit(job)
+
+        logger.info(
+            "Fetching a list of CustomContainerTrainingJobs from VertexAI server"
+        )
+        for job in self.client.CustomContainerTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info(
+            "Fetching a list of CustomPythonPackageTrainingJob from VertexAI server"
+        )
+        for job in self.client.CustomPythonPackageTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info("Fetching a list of AutoMLTabularTrainingJobs from VertexAI server")
+        for job in self.client.AutoMLTabularTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info("Fetching a list of AutoMLTextTrainingJobs from VertexAI server")
+        for job in self.client.AutoMLTextTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info("Fetching a list of AutoMLImageTrainingJobs from VertexAI server")
+        for job in self.client.AutoMLImageTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info("Fetching a list of AutoMLVideoTrainingJobs from VertexAI server")
+        for job in self.client.AutoMLVideoTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+        logger.info(
+            "Fetching a list of AutoMLForecastingTrainingJobs from VertexAI server"
+        )
+        for job in self.client.AutoMLForecastingTrainingJob.list():
+            yield from self._get_training_job_workunit(job)
+
+    def _get_training_job_workunit(
+        self, job: VertexAiResourceNoun
+    ) -> Iterable[MetadataWorkUnit]:
+        yield from self._get_data_process_properties_workunit(job)
+        yield from self._get_job_output_workunit(job)
+        yield from self._get_job_input_workunit(job)
+
+    def _get_ml_group_workunit(
+        self,
+        model: Model,
+    ) -> Iterable[MetadataWorkUnit]:
+        """
+        Generate an MLModelGroup work unit for a VertexAI  Model.
+        """
+        ml_model_group_urn = self._make_ml_model_group_urn(model)
+
+        mcp = MetadataChangeProposalWrapper(
+            entityUrn=ml_model_group_urn,
+            aspect=MLModelGroupPropertiesClass(
+                name=self._make_vertexai_model_group_name(model.name),
+                description=model.description,
+                created=TimeStampClass(time=int(model.create_time.timestamp() * 1000))
+                if model.create_time
+                else None,
+                lastModified=TimeStampClass(
+                    time=int(model.update_time.timestamp() * 1000)
+                )
+                if model.update_time
+                else None,
+                customProperties={"displayName": model.display_name},
+            ),
+        )
+
+        yield from auto_workunit([mcp])
```

Comment:
this method should just return an MCP

the `auto_workunit` call can happen in the caller


this applies throughout the code - it seems like we're calling auto_workunit too many times